### PR TITLE
fix: qualify a sandbox url via cname lookup

### PR DESCRIPTION
**Pull request by peternhale**
@W-8956222@ Mofify jwtAudienceUrl to query cns cname for sandbox qualification

### CODE REVIEW

File: src/authInfo.ts [hunk 1]
```diff
@@ -214,18 +243,11 @@ function getJwtAudienceUrl(options: OAuth2Options & { createdOrgInstance?: strin
...
-  const loginUrlLowercased = options.loginUrl?.toLowerCase();
-  if (
-    createdOrgInstance.startsWith('cs') ||
-    createdOrgInstance.endsWith('s') ||
-    loginUrlLowercased?.includes('sandbox.my.salesforce.com') || // enhanced domains >= 230
-    loginUrlLowercased?.match(/(cs[0-9]+(\.my|)\.salesforce\.com)/g) || // my domains on CS instance OR CS instance without my domain
-    loginUrlLowercased?.match(/([a-z]{3}[0-9]+s\.sfdc-.+\.salesforce\.com)/g) || // falcon sandbox ex: usa2s.sfdc-whatever.salesforce.com
-    (options.loginUrl && urlParse(options.loginUrl).hostname === 'test.salesforce.com')
-  ) {
+  if (await resolvesToSandbox(options)) {
```
Comment (peternhale) [inline src/authInfo.ts:246]: return default sandbox login url if url belongs to a sandbox

File: src/status/myDomainResolver.ts [hunk 2]
```diff
@@ -108,6 +108,17 @@ export class MyDomainResolver extends AsyncOptionalCreatable<MyDomainResolver.Op
     return ensureString(await client.subscribe());
   }
 
+  public async getCnames(): Promise<string[]> {
+    try {
+      await this.resolve();
```
Comment (peternhale) [inline src/status/myDomainResolver.ts:113]: make sure host is reachable

File: src/status/myDomainResolver.ts [hunk 3]
```diff
@@ -108,6 +108,17 @@ export class MyDomainResolver extends AsyncOptionalCreatable<MyDomainResolver.Op
...
+      return await promisify(resolveCname)(this.options.url.host);
```
Comment (peternhale) [inline src/status/myDomainResolver.ts:114]: run `resolveCname` on host and return the result []

File: src/status/myDomainResolver.ts [hunk 4]
```diff
@@ -108,6 +108,17 @@ export class MyDomainResolver extends AsyncOptionalCreatable<MyDomainResolver.Op
...
+    } catch (e) {
+      this.logger.debug(`An error occurred trying to resolve: ${this.options.url.host}`);
```
Comment (peternhale) [inline src/status/myDomainResolver.ts:116]: log the error and return empty array.

File: src/authInfo.ts [hunk 5]
```diff
@@ -203,7 +205,34 @@ export enum SfdcUrl {
...
 
-function getJwtAudienceUrl(options: OAuth2Options & { createdOrgInstance?: string }) {
+function isSandboxUrl(options: OAuth2Options & { createdOrgInstance?: string }): boolean {
+  const createdOrgInstance = getString(options, 'createdOrgInstance', '').trim().toLowerCase();
+  const loginUrl = options.loginUrl ?? '';
+  return (
+    /^cs|s$/gi.test(createdOrgInstance) ||
+    /sandbox\.my\.salesforce\.com/gi.test(loginUrl) || // enhanced domains >= 230
+    /(cs[0-9]+(\.my|)\.salesforce\.com)/gi.test(loginUrl) || // my domains on CS instance OR CS instance without my domain
+    /(cs[0-9]+(\.my|)\.salesforce\.com)/gi.test(loginUrl) || // my domains on CS instance OR CS instance without my domain
```
Comment (shetzel) [inline src/authInfo.ts:215]: duplicate line

File: src/authInfo.ts [hunk 6]
```diff
@@ -203,7 +205,34 @@ export enum SfdcUrl {
...
+}
+
+async function resolvesToSandbox(options: OAuth2Options & { createdOrgInstance?: string }): Promise<boolean> {
+  if (isSandboxUrl(options)) {
+    return true;
+  }
+  if (options.loginUrl) {
+    const myDomainResolver = await MyDomainResolver.create({ url: new URL(options.loginUrl) });
+    const cnames = await myDomainResolver.getCnames();
+    if (cnames.some((cname) => isSandboxUrl({ ...options, loginUrl: cname }))) {
```
[the remark this answers is not in the document]
Comment (shetzel) [reply src/authInfo.ts:228]: You can just return the result of `cnames.some()` right?

**Review by shetzel — CHANGES_REQUESTED** [review]
minor changes

**Review by WillieRuemmele — COMMENTED** [review]
Approved with Steve's comments

File: src/authInfo.ts [hunk 7]
```diff
@@ -203,7 +205,34 @@ export enum SfdcUrl {
   PRODUCTION = 'https://login.salesforce.com',
 }
 
-function getJwtAudienceUrl(options: OAuth2Options & { createdOrgInstance?: string }) {
+function isSandboxUrl(options: OAuth2Options & { createdOrgInstance?: string }): boolean {
```
[the remark this answers is not in the document]
Comment (mshanemc) [reply src/authInfo.ts:208]: there's a couple of these already in `util/sfdc` that might be useful outside of authinfo.  Maybe it goes there?

File: src/authInfo.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mshanemc) [reply src/authInfo.ts:228]: `array.some` is a boolean so you could just return that?

File: src/status/myDomainResolver.ts (hunk 2 above)
Comment (mshanemc) [reply src/status/myDomainResolver.ts:113]: what would happen on unresolvable if this weren't here?  wouldn't the next line still catch and return `[]` ?

File: test/unit/status/myDomainResolverTest.ts [hunk 8]
```diff
@@ -78,4 +78,69 @@ describe('myDomainResolver', () => {
       expect(e.name).to.equal('MyDomainResolverTimeoutError');
     }
   });
+
+  it('should resolve localhost', async () => {
+    const options: MyDomainResolver.Options = {
+      url: new URL('http://ghostbusters.internal.salesforce.com'),
```
Comment (mshanemc) [inline test/unit/status/myDomainResolverTest.ts:84]: is this url format still in use?  Other tests have examples with both `localhost` and a port
https://github.com/forcedotcom/sfdx-core/blob/5f17bf5635475963cbca85f6a4560d7f6aa39141/test/unit/authInfoTest.ts#L70

**Review by mshanemc — APPROVED** [review]
approved with non-blocking comments

**Review by shetzel — APPROVED** [review]

**Pull request merged**